File: iteration2/Course.java
import java.util.ArrayList;
import java.util.List;


// Course class
class Course {
    private final String COURSE_CODE;
    private final String COURSE_NAME;
    private final int COURSE_CREDIT;
    private int capacity;
    private int numberOfStudents;

    private String type;

    public int getCourseYear() {
        return COURSE_YEAR;
    }

    private final int COURSE_YEAR;
    private final CourseSection SECTION;
    private Lecturer lecturer;
    private Assistant assistant;
    private List<Course> preRequisite;

    public Lecturer getLecturer() {
        return lecturer;
    }

    public void setLecturer(Lecturer lecturer) {
        this.lecturer = lecturer;
    }

    public Assistant getAssistant() {
        return assistant;
    }

    public void setAssistant(Assistant assistant) {
        this.assistant = assistant;
    }

    public List<Course> getPreRequisite() {
        return preRequisite;
    }

    public void setPreRequisite(List<Course> preRequisite) {
        this.preRequisite = preRequisite;
    }

    // Constructor
    public Course(String courseCode, String courseName, int courseCredit, int courseYear, int section1, int section2, Lecturer lecturer, List<Course> preRequisite) {
        this.COURSE_CODE = courseCode;
        this.COURSE_NAME = courseName;
        this.COURSE_CREDIT = courseCredit;
        this.COURSE_YEAR = courseYear;
        this.SECTION = new CourseSection(section1, section2);
        this.lecturer = lecturer;
        this.preRequisite = preRequisite;
        setType("n");
    }

    public Course(String courseCode, String courseName, int courseCredit, int courseYear, int section1, int section2, Lecturer lecturer, Assistant assistant, List<Course> preRequisite) {
        this.COURSE_CODE = courseCode;
        this.COURSE_NAME = courseName;
        this.COURSE_CREDIT = courseCredit;
        this.COURSE_YEAR = courseYear;
        this.assistant = assistant;
        this.SECTION = new CourseSection(section1, section2);
        this.lecturer = lecturer;
        this.preRequisite = preRequisite;
        setType("n");
    }

    public Course(String courseCode, String courseName, int courseCredit, int courseYear, int section1, int section2, Lecturer lecturer, Assistant assistant, List<Course> preRequisite, int capacity) {
        this.COURSE_CODE = courseCode;
        this.COURSE_NAME = courseName;
        this.COURSE_CREDIT = courseCredit;
        this.COURSE_YEAR = courseYear;
        this.SECTION = new CourseSection(section1, section2);
        this.lecturer = lecturer;
        this.preRequisite = preRequisite;
        this.capacity = capacity;
        setType("n");
    }


    // Check if this student in the appropriate semester to take this course?
    public boolean checkYearMatching(int year) {
        return COURSE_YEAR <= year;
    }

    // Check if this course conflicts with any other courses section returns 0 for available course
    public String checkCourseSection(List<Course> registrationCompleteCourses, List<Course> registrationWaitingCourses, List<Course> cancelWaitingCourses) {
        for (Course course : registrationCompleteCourses) {
            if (this.equals(course)) {
                return "This course:(" + course.toString() + ")" + "already exist in completedRegistrationCourses.Try to select another course";
            }
            if (!course.getSection().compareAvailability(this.SECTION)) {
                return "This course is conflicting with the time of the course : " + course.toString() + " inside your courses that are registered!";
            }
        }
        for (Course course : registrationWaitingCourses) {
            if (this.equals(course)) {
                return "This course:(" + course.toString() + ")" + " already exist in registrationWaitingCourses.Try to select another course";
            }
            if (!course.getSection().compareAvailability(this.SECTION)) {
                return "This course is conflicting with the time of the course : " + course.toString() + " inside your courses that are waiting to be registered!";
            }
        }
        for (Course course : cancelWaitingCourses) {
            if (this.equals(course)) {
                return "This course:(" + course.toString() + ")" + " already exist in cancelWaitingCourses.Try to select another course";
            }
            if (!course.getSection().compareAvailability(this.SECTION)) {
                return "This course is conflicting with the time of the course : " + course.toString() + " inside your courses that are waiting to be canceled!";
            }
        }
        if (isFull()) {
            return "this course's capacity full!.";
        }
        return null;
    }


    public boolean equals(Course course2) {
        return course2 != null && this.COURSE_NAME.equals(course2.getCourseName());
    }

    // Check if given student can take this course, according to prerequisite and given completed courses
    public boolean checkPreRequisite(List<Course> completedCourses, List<Grade> gradeList) {
        if (preRequisite == null) {
            return true;
        }
        if (preRequisite.isEmpty()) {
            return true;
        }
        List<Course> copy = new ArrayList<>(preRequisite);

        for (Course course : completedCourses) {
            for (int i = 0; i < copy.size(); i++) {
                Course preRequisiteCurrent = copy.get(i);
                if (course.equals(preRequisiteCurrent) && gradeList.get(i).isPassedGrade()) {
                    copy.remove(preRequisiteCurrent);
                }
            }
            if (copy.isEmpty()) {
                return true;
            }
        }

        return false;
    }

    // Returns course information, code and name
    public String toString() {
        return COURSE_CODE + " " + COURSE_NAME;
    }

    public String courseStatus() {
        return "NumberOfStudent: " + numberOfStudents + " " + "Capacity: " + capacity;
    }

    public int getCourseCredit() {
        return COURSE_CREDIT;
    }

    public String getCourseCode() {
        return COURSE_CODE;
    }

    public CourseSection getSection() {
        return SECTION;
    }

    public String getCourseName() {
        return COURSE_NAME;
    }

    public void increaseStudentNumber() {
        numberOfStudents++;
    }

    public boolean isFull() {
        return numberOfStudents >= capacity;
    }

    public void decreaseStudentNumber() {
        numberOfStudents--;
    }

    public void setType(String type) {
        switch (type) {
            case "t" -> this.type = "t";
            case "nt" -> this.type = "nt";
            case "n" -> this.type = "n";
            default -> System.out.println("Type does not exist! instead try:\nt, nt or n.");
        }
    }
}

File: iteration3/Course.py
import logging

from CourseSection import CourseSection


# This class only has one __init__ !!
# use this with None and 0 if you wish to initialize this class without assistant or capacity!!

class Course:
    __course_code, __course_name, __type = "", "", ""
    __lecturer, __assistant, __section = None, None, None
    __course_credit, __course_year, __capacity, __number_of_students = int(0)
    __prerequisite: list = []

    # Just use this constructor with assistant and capacity, if assistant and capacity are unused give None and 0!!
    def __init__(self, course_code, course_name, course_credit, course_year, section1, section2, lecturer, prerequisite,
                 assistant, capacity):
        self.__course_code = course_code
        self.__course_name = course_name
        self.__course_credit = course_credit
        self.__course_year = course_year
        self.__section = CourseSection(section1, section2)
        self.__lecturer = lecturer
        self.__prerequisite = prerequisite
        self.__capacity = capacity
        self.__assistant = assistant
        self.set_type("n")

    def check_course_section(self, registration_complete_courses: list,
                             registration_waiting_courses: list,
                             cancel_waiting_courses: list):
        for course in registration_complete_courses:
            if not isinstance(course, Course):
                logging.error("course object in registration_complete_courses not an instance of Course class!")
            else:
                if self.equals(course):
                    return ("This course:(" + course.get_course_string() +
                            ")" + "already exist in completedRegistrationCourses.Try to select another course")
                if not course.get_course_section().compare_availability(self.__section):
                    return ("This course is conflicting with the time of the course : " + course.get_course_string()
                            + " inside your courses that are registered!")
        for course in registration_waiting_courses:
            if not isinstance(course, Course):
                logging.error("course object in registration_waiting_courses not an instance of Course class!")
            else:
                if self.equals(course):
                    return ("This course:(" + course.get_course_string() +
                            ")" + "already exist in registrationWaitingCourses.Try to select another course")
                if not course.get_course_section().compare_availability(self.__section):
                    return ("This course is conflicting with the time of the course : " + course.get_course_string()
                            + " inside your courses that are waiting to be registered!")
        for course in cancel_waiting_courses:
            if not isinstance(course, Course):
                logging.error("course object in cancel_waiting_courses not an instance of Course class!")
            else:
                if self.equals(course):
                    return ("This course:(" + course.get_course_string() +
                            ")" + "already exist in cancelWaitingCourses.Try to select another course")
                if not course.get_course_section().compare_availability(self.__section):
                    return ("This course is conflicting with the time of the course : " + course.get_course_string()
                            + " inside your courses that are waiting to be cancelled!")
        if self.is_full():
            return "this course's capacity full!."
        return None


    def check_technical_elective_count(self, registration_complete_courses: list,
                                       registration_waiting_courses: list,
                                       cancel_waiting_courses: list):
        amount: int = 0
        threshold: int = 5
        for course in registration_waiting_courses:
            if not isinstance(course, Course):
                logging.error("course object in registration_waiting_courses not an instance of Course class!")
            elif course.__type == "t":
                amount += 1
        for course in registration_complete_courses:
            if not isinstance(course, Course):
                logging.error("course object in registration_complete_courses not an instance of Course class!")
            elif course.__type == "t":
                amount += 1
        for course in cancel_waiting_courses:
            if not isinstance(course, Course):
                logging.error("course object in cancel_waiting_courses not an instance of Course class!")
            elif course.__type == "t":
                amount += 1
        if amount >= threshold:
            return ("This course exceeds the Technical elective course count."
                    " Try cancelling another Technical elective or waiting for the cancellation to be accepted.")
        else:
            return None

    def check_non_technical_elective_count(self, registration_complete_courses: list,
                                           registration_waiting_courses: list,
                                           cancel_waiting_courses: list):
        amount: int = 0
        threshold: int = 3
        for course in registration_waiting_courses:
            if not isinstance(course, Course):
                logging.error("course object in registration_waiting_courses not an instance of Course class!")
            elif course.__type == "nt":
                amount += 1
        for course in registration_complete_courses:
            if not isinstance(course, Course):
                logging.error("course object in registration_complete_courses not an instance of Course class!")
            elif course.__type == "nt":
                amount += 1
        for course in cancel_waiting_courses:
            if not isinstance(course, Course):
                logging.error("course object in cancel_waiting_courses not an instance of Course class!")
            elif course.__type == "nt":
                amount += 1
        if amount >= threshold:
            return ("This course exceeds the Non-Technical elective course count."
                    " Try cancelling another Non-Technical elective or waiting for the cancellation to be accepted.")
        else:
            return None

    def check_prerequisite(self, completed_courses: list, grade_list: list) -> bool:
        if isinstance(self.__prerequisite, Course):
            if self.__prerequisite is None:
                return True
            if not self.__prerequisite:
                return True

        copy = self.__prerequisite

        for course in completed_courses:
            for i in range(len(copy)):
                prerequisite_current: Course = copy[i]
                if course.equals(prerequisite_current) and grade_list[i].is_passed_grade():
                    copy.remove(prerequisite_current)
            if not copy:
                return True
        return False

    def check_year_matching(self, year: int) -> bool:
        return self.__course_year <= year

    def equals(self, c2) -> bool:
        if isinstance(c2, Course):
            return (c2 is not None and
                    self.__course_name == c2.get_course_name())
        else:
            return False

    def set_type(self, ty: str):
        if ty == "t":
            self.__type = "t"
        elif ty == "nt":
            self.__type = "nt"
        elif ty == "n":
            self.__type = "n"
        else:
            logging.error("Type does not exist! instead try:\nt, nt or n.")

    def get_course_name(self):
        return self.__course_name

    def get_course_type(self):
        return self.__type

    def get_course_credit(self):
        return self.__course_credit

    def get_course_code(self):
        return self.__course_code

    def get_course_section(self):
        return self.__section

    def increment_student_amount(self):
        self.__number_of_students += self.__number_of_students

    def decrement_student_amount(self):
        self.__number_of_students -= self.__number_of_students

    def is_full(self):
        return self.__number_of_students >= self.__capacity

    def get_course_string(self):
        return self.__course_code + " " + self.__course_name

    def course_status(self):
        return "NumberOfStudent: " + self.__number_of_students + " " + "Capacity: " + self.__capacity
